Implement Disklet sync

samholmes opened the pull request:
### CHANGELOG

Does this branch warrant an entry to the CHANGELOG?

- [x] Yes
- [ ] No

### Dependencies

<!-- Replace line with PRs which this PR depends if any --> none

### Description

<!-- Describe your changes textually and/or pictorially if necessary --> none


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1209070963485246
  - https://app.asana.com/0/0/1212383879923573

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Introduces Disklet-backed synchronization and associated abstractions.
> 
> - Adds `SyncClient.syncRepo(disklet, syncKey, lastHash)` to interlace `changes/` and `deleted/`, apply server changes to `data/`, and respect `maxChangesPerSync` (default 100)
> - New `makeSyncedDisklet` and `SyncedDisklet` to transparently stage writes/deletes and expose `sync()`; updates exports and Flow/types (`SyncStatus`, `SyncResult`)
> - Enhances encoding: adds `asBase16`/`asBase64`, updates `EdgeBox` cleaners and exports `wasEdgeBox`
> - Replaces info-client usage with inline `defaultEdgeServers`; supports `edgeServers.syncServers`
> - Build: add `disklet` as peer dep, include peer deps in Rollup externals; bump version to `0.2.9-1`
> - Tests: comprehensive `syncRepo` component tests and mock sync server
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 103e45f2d275a280e4d9391b36e15f8351fef116. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

samholmes committed:
Remove InfoClient usage for SyncClient instances

samholmes committed:
v0.2.9-1

samholmes committed:
Implement Disklet synchronization

samholmes committed:
Merge branch '0.2.9-1'

swansontec commented on src/client/sync-client.ts:
```diff
@@ -182,6 +210,94 @@ export function makeSyncClient(opts: SyncClientOptions = {}): SyncClient {
       throw error
+    },
+
+    async syncRepo(disklet, syncKey, lastHash) {
```
This function seems reasonable. However, this should not be public API. Instead, we want to export a `makeSyncedDisklet(parentPath: Disklet) => Disklet & { sync() => Promise<string[]> }`.

You do `const myDisklet = makeSyncedDisklet(...)`, and then you can do `myDisklet.getText`, `myDisklet.delete`, `myDisklet.setText`, and so forth. This library automatically routes the changes to the right places (/data, /changes, /deleted). Finally, you can do `myDisklet.sync()` to sync with the server.

This keeps the details fully black-boxed. The contents of the "status.json", the layout of the sub-folders (/data, /changes, /deleted), and the details of the sync algorithm are fully abstracted. Remember when we built the change server? We built the "legacy" API compatible with the existing core, and we also defined the "new" API with faster sync. We never implemented the client side of the "new" API, but if we wanted to do that, we would want edge-sync-client to be "black-boxed" so the change can be seamless. If the core has to "know" the format of the "status.json" or the general on-disk layout of the repo, we have failed.

samholmes committed:
fixup! Implement Disklet synchronization

This function seems reasonable. However, this should not be public API. Instead, we want to export a `makeSyncedDisklet(parentPath: Disklet) => Disklet & { sync() => Promise<string[]> }`.

You do `const myDisklet = makeSyncedDisklet(...)`, and then you can do `myDisklet.getText`, `myDisklet.delete`, `myDisklet.setText`, and so forth. This library automatically routes the changes to the right places (/data, /changes, /deleted). Finally, you can do `myDisklet.sync()` to sync with the server.

This keeps the details fully black-boxed. The contents of the "status.json", the layout of the sub-folders (/data, /changes, /deleted), and the details of the sync algorithm are fully abstracted. Remember when we built the change server? We built the "legacy" API compatible with the existing core, and we also defined the "new" API with faster sync. We never implemented the client side of the "new" API, but if we wanted to do that, we would want edge-sync-client to be "black-boxed" so the change can be seamless. If the core has to "know" the format of the "status.json" or the general on-disk layout of the repo, we have failed.

Addresses: https://github.com/EdgeApp/edge-sync-client/pull/19#discussion_r2608550317
